control/prometheus.py

```diff
@@ -0,0 +1,321 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: pcuzner@ibm.com
+#
+
+import os
+import time
+import logging
+import threading
+import inspect
+import spdk.rpc as rpc
+
+from prometheus_client.core import REGISTRY, GaugeMetricFamily, CounterMetricFamily, InfoMetricFamily
+from prometheus_client import start_http_server, GC_COLLECTOR
+from typing import NamedTuple
+from functools import wraps
+
+COLLECTION_ELAPSED_WARNING = 0.8   # Percentage of the refresh interval before a warning message is issued
+REGISTRY.unregister(GC_COLLECTOR)  # Turn off garbage collector metrics
+
+logger = logging.getLogger(__name__)
+
+
+class RBD(NamedTuple):
+    pool: str
+    namespace: str
+    image: str
+
+
+def ttl(method):
+    @wraps(method)
+    def wrapped(self, *args, **kwargs):
+        assert inspect.isgeneratorfunction(method)
+        with self.lock:
+            now = time.time()
+            if now - self.last_obs >= self.interval:
+                self.last_obs = now
+                self.metrics_cache.clear()
+                for metric in method(self, *args, **kwargs):
+                    self.metrics_cache.append(metric)
+                    yield metric
+            else:
+                logger.debug('Returning content from cache')
+                for metric in self.metrics_cache:
+                    yield metric
```

Comment:
Just a comment: a `yield` inside a `with self.lock` section might ([not 100% sure](https://stackoverflow.com/questions/685046/is-it-safe-to-yield-from-within-a-with-block-in-python-and-why)) result in the lock held for periods longer than intended and potential deadlocks.

My question is: do we really need a yield there? Can we simply return a plain iterable, like a list?
```suggestion
    def wrapped(self, *args, **kwargs):
        assert inspect.isgeneratorfunction(method)
        with self.lock:
            now = time.time()
            if now - self.last_obs >= self.interval:
                self.last_obs = now
                self.metrics_cache = list(method(self, *args, **kwargs))
            return self.metrics_cache
```
Not sure if that'll work, but at least we're returning all metrics at once.

Reply:
I've always used yield which, looking at the client code itself, seems to be the general convention used. I'd prefer to keep things vanilla. If this wasn't a python project for example, the values would be pushed to the channel (pipe) as and when they're available - which again aligns to the yield pattern.

Reply:
@epuertat if you're happy with my reply, please resolve the thread.

Reply:
Ok, no big issue to me!

I would just ask you if you can try the following:
* Add a sleep(20) in the middle the `collect()` method,
* Scrape the HTTP endpoint (cold cache) and immediately issue another request.

My guess is that the second scrape will block for 20+ secs (or the network timeout, whichever happens first). This doesn't really look like an invalid behaviour, agree, but perhaps we could specify a shorter time-out to the lock acquisition and return a 503 with a [`Retry-after` directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After)?

Unfortunately the context manager approach (`with lock`) [doesn't support specifying that](https://stackoverflow.com/questions/16740104/python-lock-with-statement-and-timeout).

In any case, take this just as an observation for the case we ever see the metric gathering underperforming. Nothing really blocking.